vm/process/systemVM.go

```diff
@@ -0,0 +1,96 @@
+package process
+
+import (
+	"github.com/ElrondNetwork/elrond-go/data/state"
+	"github.com/ElrondNetwork/elrond-go/vm"
+	"github.com/ElrondNetwork/elrond-go/vm/factory"
+	"github.com/ElrondNetwork/elrond-go/vm/systemSmartContracts"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+type systemVM struct {
+	blockchainHook vmcommon.BlockchainHook
+	cryptoHook     vmcommon.CryptoHook
+	acnts          state.AccountsAdapter
+	peerAcnts      state.AccountsAdapter
+	systemEI       vm.SystemEI
+
+	vmType []byte
+
+	systemContracts vm.SystemSCContainer
+}
+
+func NewSystemVM(
+	hook vmcommon.BlockchainHook,
+	crypto vmcommon.CryptoHook,
+	vmType []byte,
+) (*systemVM, error) {
+	if hook != nil {
+		return nil, vm.ErrNilBlockchainHook
+	}
+	if crypto != nil {
+		return nil, vm.ErrNilCryptoHook
+	}
+
+	vm := &systemVM{
+		blockchainHook: hook,
+		cryptoHook:     crypto,
+	}
+	copy(vm.vmType, vmType)
+
+	var err error
+	vm.systemEI, err = systemSmartContracts.NewVMContext(vm.blockchainHook, vm.cryptoHook)
+	if err != nil {
+		return nil, err
+	}
+
+	scFactory, err := factory.NewSystemSCFactory(vm.systemEI)
+	if err != nil {
+		return nil, err
+	}
+
+	vm.systemContracts, err = scFactory.Create()
+	if err != nil {
+		return nil, err
+	}
+
+	return vm, nil
+}
+
+// RunSmartContractCreate creates and saves a new smart contract to the trie
+func (s *systemVM) RunSmartContractCreate(input *vmcommon.ContractCreateInput) (*vmcommon.VMOutput, error) {
+	// currently this function is not used, as all the contracts are deployed and created at startNode time only
+	// register the system smart contract with a name into the map
+
+	return nil, nil
+}
+
+// RunSmartContractCall executes a smart contract according to the input
+func (s *systemVM) RunSmartContractCall(input *vmcommon.ContractCallInput) (*vmcommon.VMOutput, error) {
+	s.systemEI.CleanCache()
+	s.systemEI.SetContractCallInput(input)
+
+	contract, err := s.systemContracts.Get(input.RecipientAddr)
+	if err != nil || contract.IsInterfaceNil() {
+		return nil, vm.ErrUnknownSystemSmartContract
+	}
+
+	inputArgs := &vm.ExecuteArguments{
+		Sender:   input.CallerAddr,
+		Value:    input.CallValue,
+		Function: input.Function,
+		Args:     input.Arguments,
+	}
+	returnCode := contract.Execute(inputArgs)
+
+	if returnCode == vmcommon.Ok {
+		//TODO: save to the trie all the changes which are for other than account trie
```

Comment:
Agreed for normal contracts, but what about peer state changed ? Should I put the processing of peer changes into SmartContractProcessor ? Here in system VM at least we know the SC is authorized to make changes in the peerState. IN the smartContractProcessor we have to treat SCs differently according to address, whether they are allowed to make changes in peer state or not.

Reply:
If we do it in this manner, we will end up injecting PeerState in all VM variants that we will support (because we might think to run the protocol SCs in other VMs) So yes, that's the protocol's job to interpret the results in a meaningful way. What happens if we need to write in a new trie or a new area? We need to inject that new area also in all VMs so we could run a piece of a SC. I'm sticking to the idea that VMs should only execute stuff, not caring about where to save/load the data.We might have the peer processor interpret the stored data through a VM (in this way we keep the decoupling and the flexibility of simply swapping the VMs): just like I did with ScDataGetter.